test/test_assign.py

```diff
@@ -107,6 +126,69 @@ def test_assign_diamond_alt(self):
     new = a + times_a
     np.testing.assert_allclose(new.numpy(), 8)
 
+  @unittest.expectedFailure
+  def test_assign_fork_ooo(self):
+    a = Tensor.ones(4).contiguous().realize()
+    times_a = a*3
+    a.assign(Tensor.full((4,), 2.).contiguous())
+    times_a = times_a * 5  # should be toposorted before the assign
+    Tensor.corealize([a, times_a])
+    np.testing.assert_allclose(a.numpy(), 2)
+    np.testing.assert_allclose(times_a.numpy(), 15)
+
+  @unittest.expectedFailure
+  def test_assign_fork_multirealize(self):
+    a = Tensor.ones(4).contiguous().realize()
+    times_a = a*3
+    a.assign(Tensor.full((4,), 2.).contiguous()).realize()
+    with self.assertRaises(Exception):
+      # naively, times_a would be calculated with the new assigned a now :(
+      times_a.realize()
+    np.testing.assert_allclose(times_a.numpy(), 3)
+    np.testing.assert_allclose(a.numpy(), 2)
+
+  def test_assign_twice(self):
+    # eg if we do multiple steps in 1 schedule
+    a = Tensor.ones(4).contiguous().realize()
+    a += a * 2
+    a += a * 4
+    np.testing.assert_allclose(a.numpy(), 15)
+
+  def test_assign_fib(self):
+    # test that things happen in the correct order
+    a = Tensor.ones(4).contiguous().realize()
+    b = Tensor.ones(4).contiguous().realize()
+    a += b
+    b += a
+    a += b
+    b += a
+    Tensor.corealize([a, b])
+    np.testing.assert_allclose(a.numpy(), 5)
+    np.testing.assert_allclose(b.numpy(), 8)
+
+  def test_assign_cycle(self):
+    a = Tensor.ones(4).contiguous().realize()
+    b = Tensor.ones(4).contiguous().realize()
+    c = a + 1
+    a.assign(b + 1)
+    b.assign(c)
+    with self.assertRaises(Exception):
+      # there is a cycle in the dag for this one.
```

Comment:
I don't think this should be a cycle, it should keep the first generation a around.